[Azure] Check & Raise TagsCreationForbidden Exception

AlekhyaYalla opened the pull request:
**Issue:** 
Today, while creating pull requests(PR), labels are passed for Azure [create_pull_request method](https://github.com/dependabot/dependabot-core/blob/d734dc6e4fe0f7903ee97e0079ac9f5362e408c1/common/lib/dependabot/pull_request_creator/azure.rb#L81) & created by default.(may it be default/security etc)
In one of our repo, the user who tries to create a dependency update PR, has permissions to create a PR in that repo, but not for the tags. So the PR creation fails, as dependabot-core tries to create tags first and then PR. As the tags creation step is failing with 403 forbidden errors, PR creation also failed. 

**Fix:**
- As Tags creation is failed and the same has been clearly mentioned in the error message, instead of raising a regular 403: Forbidden exception, we can raise TagsCreationForbidden exception. For this, we can validate the error code in the message & raise the same. This exception is only seen in Azure, so handled in the Azure only.

jurre commented:
> So instead of creating a new object of pull_request_creator, I thought azure_creator can be created based on a new field/param that would talk about if this creation need labels or not.

I'm not sure that I really understand the difference or at least why one is better over the other?

> And also, I thought custom_labels is meant for passing if the user has any custom labels for the PR. Using it in this case may sound a little ambigous

While I agree that the naming is not as clear as it could be, the functionality is already there to create PRs without any labels, I think having both the option of passing an empty array to `custom_labels` _and_ also having the `labels_required` option leaves us worse-off?

AlekhyaYalla commented:
Hi @jurre 
Based on the discussion we had on chat, concluding the behaviour to attempt labels creation in PullRequestCreator can be achieved by passing `[]` to the customLabels. Removed that part.
Added `TagsCreationForbidden` in the azure client as it is a valid case.

AlekhyaYalla commented on common/lib/dependabot/pull_request_creator.rb:
```diff
@@ -60,7 +60,7 @@ def initialize(source:, base_commit:, dependencies:, files:, credentials:,
                    github_redirection_service: DEFAULT_GITHUB_REDIRECTION_SERVICE,
                    custom_headers: nil, require_up_to_date_base: false,
-                   provider_metadata: {}, message: nil)
+                   provider_metadata: {}, message: nil, labels_required: true)
```
Removed the param `labels_required`

jurre commented on common/spec/dependabot/clients/azure_spec.rb:
```diff
@@ -181,6 +181,39 @@
+      before do
+        stub_request(:post, pull_request_url).
+          with(basic_auth: [username, password]).
+          to_return(status: 403, body: JSON.unparse({ message: "TF401289: The current user does not have permissions to create tags" }))
```
```suggestion
          to_return(
            status: 403,
            body: { message: "TF401289: The current user does not have permissions to create tags" }.to_json
          )
```

The linter is currently failing because this line is too long, and the `.to_json` is a bit more in line with the rest of the codebase.

AlekhyaYalla replied:
Sure will be addressing this.

jurre commented on common/lib/dependabot/clients/azure.rb:
```diff
@@ -310,6 +317,13 @@ def truncate_pr_description(pr_description)
+        return if response.body.empty?
+
+        message = JSON.parse(response.body).fetch("message", nil)
+        message&.include?("TF401289: The current user does not have permissions to create tags")
```
Do you think it would be sufficient to just match on `TF401289`, that way if the copy in that error ever changes we don't need to patch this code, and it seems like this is a code that specifically represents this tags permission error?

AlekhyaYalla replied:
Seems like that works, as `TF401289` is the message code for the permission tags issue. Will update the fix

AlekhyaYalla commented:
> @AlekhyaYalla would you mind updating the PR title / description to reflect the changes we landed on in this PR, that way, if we ever go back to figure out why this code was changed, the PR will reflect the relevant background info

Updated @jurre My apologies for not updating.
